[APM] Optimize idx calls to native optional chaining

ogupte opened the pull request:
Closes #32559.
Adds new kibana package '@kbn/babel-plugin-apm-idx' based on 'babel-plugin-idx' to work with the APM implementation of the deep object property helper function. Configures it as a babel plugin in the common babel preset.

Example:
The plugin transpiles this:
```javascript
import { idx } from '@kbn/elastic-idx';
idx(transaction, _ => _.context.page.url)
```
to this:
```javascript
transaction != null && transaction.context != null && transaction.context.page != null
    ? transaction.context.page.url
    : undefined;
```
removing the import of the idx module, reducing bundle size.

ogupte committed:
[APM] Add new kibana package '@kbn/babel-plugin-apm-idx' based on
'babel-plugin-idx' to work with the APM implementation of the deep
object property helper function. Configure it as a babel plugin in the
common babel preset.

ogupte commented on packages/kbn-babel-plugin-apm-idx/index.js:
```diff
@@ -0,0 +1,251 @@
+/*
```
Diff against the original work (https://github.com/facebookincubator/idx/blob/462c7ecd3ca717503c35606be0e7f355715ddf99/packages/babel-plugin-idx/src/babel-plugin-idx.js):
```
diff --git a/./babel-plugin-idx.js b/./kbn-babel-plugin-apm-idx.js
index 81f4f65..d37eb28 100644
--- a/./babel-plugin-idx.js
+++ b/./kbn-babel-plugin-apm-idx.js
@@ -1,3 +1,28 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * The following is the original license notice in the source
+ * code from which this code is based.
+ * (https://github.com/facebookincubator/idx/blob/462c7ecd3ca717503c35606be0e7f355715ddf99/packages/babel-plugin-idx/src/babel-plugin-idx.js)
+ */
+
 /**
  * Copyright (c) 2013-present, Facebook, Inc.
  *
@@ -7,8 +32,9 @@
  * @format
  */
 
-'use strict'; // eslint-disable-line strict
+/* eslint strict: 0, new-cap: 0 */
 
+'use strict';
 module.exports = context => {
   const t = context.types;
 
@@ -17,36 +43,33 @@ module.exports = context => {
   function checkIdxArguments(file, node) {
     const args = node.arguments;
     if (args.length !== 2) {
-      throw file.buildCodeFrameError(
-        node,
-        'The `idx` function takes exactly two arguments.',
-      );
+      throw file.buildCodeFrameError(node, 'The `idx` function takes exactly two arguments.');
     }
     const arrowFunction = args[1];
     if (!t.isArrowFunctionExpression(arrowFunction)) {
       throw file.buildCodeFrameError(
         arrowFunction,
-        'The second argument supplied to `idx` must be an arrow function.',
+        'The second argument supplied to `idx` must be an arrow function.'
       );
     }
     if (!t.isExpression(arrowFunction.body)) {
       throw file.buildCodeFrameError(
         arrowFunction.body,
         'The body of the arrow function supplied to `idx` must be a single ' +
-          'expression (without curly braces).',
+          'expression (without curly braces).'
       );
     }
     if (arrowFunction.params.length !== 1) {
       throw file.buildCodeFrameError(
         arrowFunction.params[2] || arrowFunction,
-        'The arrow function supplied to `idx` must take exactly one parameter.',
+        'The arrow function supplied to `idx` must take exactly one parameter.'
       );
     }
     const input = arrowFunction.params[0];
     if (!t.isIdentifier(input)) {
       throw file.buildCodeFrameError(
         arrowFunction.params[0],
-        'The parameter supplied to `idx` must be an identifier.',
+        'The parameter supplied to `idx` must be an identifier.'
       );
     }
   }
@@ -55,10 +78,7 @@ module.exports = context => {
     if (t.isImportDeclaration(node)) {
       // E.g. `import '...'`
       if (node.specifiers.length === 0) {
-        throw file.buildCodeFrameError(
-          node,
-          'The idx import must have a value.',
-        );
+        throw file.buildCodeFrameError(node, 'The idx import must have a value.');
       }
       // E.g. `import A, {B} from '...'`
       //      `import A, * as B from '...'`
@@ -66,16 +86,7 @@ module.exports = context => {
       if (node.specifiers.length > 1) {
         throw file.buildCodeFrameError(
           node.specifiers[1],
-          'The idx import must be a single specifier.',
-        );
-      }
-      // `import {default as idx} from '...'` or `import idx from '...'` are ok.
-      // `import idx, * as idx2 from '...'` is not ok but would've been caught
-      // above.
-      if (!t.isSpecifierDefault(node.specifiers[0])) {
-        throw file.buildCodeFrameError(
-          node.specifiers[0],
-          'The idx import must be a default import.',
+          'The idx import must be a single specifier.'
         );
       }
       // `importKind` is not a property unless flow syntax is enabled.
@@ -91,17 +102,14 @@ module.exports = context => {
         node.specifiers[0].importKind === 'type' ||
         node.specifiers[0].importKind === 'typeof'
       ) {
-        throw file.buildCodeFrameError(
-          node,
-          'The idx import must be a value import.',
-        );
+        throw file.buildCodeFrameError(node, 'The idx import must be a value import.');
       }
     } else if (t.isVariableDeclarator(node)) {
       // E.g. var {idx} or var [idx]
       if (!t.isIdentifier(node.id)) {
         throw file.buildCodeFrameError(
           node.specifiers[0],
-          'The idx declaration must be an identifier.',
+          'The idx declaration must be an identifier.'
         );
       }
     }
@@ -110,13 +118,9 @@ module.exports = context => {
   function makeCondition(node, state, inside) {
     if (inside) {
       return t.ConditionalExpression(
-        t.BinaryExpression(
-          '!=',
-          t.AssignmentExpression('=', state.temp, node),
-          t.NullLiteral(),
-        ),
+        t.BinaryExpression('!=', t.AssignmentExpression('=', state.temp, node), t.NullLiteral()),
         inside,
-        state.temp,
+        state.temp
       );
     } else {
       return node;
@@ -128,25 +132,21 @@ module.exports = context => {
       return makeChain(
         node.object,
         state,
-        makeCondition(
-          t.MemberExpression(state.temp, node.property, node.computed),
-          state,
-          inside,
-        ),
+        makeCondition(t.MemberExpression(state.temp, node.property, node.computed), state, inside)
       );
     } else if (t.isIdentifier(node)) {
       if (node.name !== state.base.name) {
         throw state.file.buildCodeFrameError(
           node,
           'The parameter of the arrow function supplied to `idx` must match ' +
-            'the base of the body expression.',
+            'the base of the body expression.'
         );
       }
       return makeCondition(state.input, state, inside);
     } else {
       throw state.file.buildCodeFrameError(
         node,
-        'idx callbacks may only access properties on the callback parameter.',
+        'idx callbacks may only access properties on the callback parameter.'
       );
     }
   }
@@ -163,21 +163,27 @@ module.exports = context => {
     });
     path.replaceWith(replacement);
     // Hoist to the top if it's an async method.
-    if (path.scope.path.isClassMethod({async: true})) {
-      path.scope.push({id: temp, _blockHoist: 3});
+    if (path.scope.path.isClassMethod({ async: true })) {
+      path.scope.push({ id: temp, _blockHoist: 3 });
     } else {
-      path.scope.push({id: temp});
+      path.scope.push({ id: temp });
     }
   }
 
   function isIdxImportOrRequire(node, name) {
     if (t.isImportDeclaration(node)) {
-      return t.isStringLiteral(node.source, {value: name});
+      if (name instanceof RegExp) {
+        return name.test(node.source.value);
+      } else {
+        return t.isStringLiteral(node.source, { value: name });
+      }
     } else if (t.isVariableDeclarator(node)) {
       return (
         t.isCallExpression(node.init) &&
-        t.isIdentifier(node.init.callee, {name: 'require'}) &&
-        t.isLiteral(node.init.arguments[0], {value: name})
+        t.isIdentifier(node.init.callee, { name: 'require' }) &&
+        (name instanceof RegExp
+          ? name.test(node.init.arguments[0].value)
+          : t.isLiteral(node.init.arguments[0], { value: name }))
       );
     } else {
       return false;
@@ -198,10 +204,7 @@ module.exports = context => {
       const idxBinding = path.scope.getOwnBinding(bindingName);
 
       idxBinding.constantViolations.forEach(refPath => {
-        throw state.file.buildCodeFrameError(
-          refPath.node,
-          '`idx` cannot be redefined.',
-        );
+        throw state.file.buildCodeFrameError(refPath.node, '`idx` cannot be redefined.');
       });
 
       let didTransform = false;
@@ -239,7 +242,7 @@ module.exports = context => {
           // "babel-plugin-transform-async-to-generator", will convert arrow
           // functions inside async functions into regular functions. So we do
           // our transformation before any one else interferes.
-          const newState = {file: state.file, importName};
+          const newState = { file: state.file, importName };
           path.traverse(declareVisitor, newState);
         }
       },
```

ogupte commented on packages/kbn-babel-preset/common_preset.js:
```diff
@@ -32,6 +29,12 @@ module.exports = {
+    [
+      require.resolve('@kbn/babel-plugin-apm-idx'),
+      {
+        importName: /\/common\/idx$/,
```
importName now supports a RegExp pattern to test Import Specifiers in order to allow our use case, i.e. relative path imports of `*/common/idx.ts`

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8968/)

sqren commented:
My intention for forking `idx` was to solve https://github.com/facebookincubator/idx/pull/71 so that
`idx({context: { page: null }}, _ => _.context.page.url);` returns `undefined` instead of `null`.

How is it currently different from `idx`?

sqren commented:
To elaborate on the above:

Input
```js
idx(transaction, _ => _.context.page.url)
```

Should be transformed to something like:
```js
transaction != null && transaction.context != null && transaction.context.page != null ? transaction.context.page.url : undefined
```

This should return `undefined`
```js
const transaction = {context: { page: null}};
idx(transaction, _ => _.context.page.url) // undefined
```

This should return `null`
```js
const transaction = {context: { page: null}};
idx(transaction, _ => _.context.page) // null
```

ogupte committed:
[APM] modified the expanded output code to handle null checking correction for nested properties and added tests.

ogupte commented on packages/kbn-babel-plugin-apm-idx/index.js:
```diff
@@ -0,0 +1,306 @@
+      }
+    }
+  }
+
```
New code in lines 118 - 215 significantly changes the output of the transpiled code from the original.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9576/)

ogupte committed:
[APM] use same versions of babel and jest as core kibana. fixes broken tests.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9736/)

tylersmalley commented on packages/kbn-babel-plugin-apm-idx/index.js:
```diff
@@ -0,0 +1,306 @@
+ */
+
+/*
+ * The following is the original license notice in the source
```
Need to add `@notice` to this in order to be included in our NOTICE.txt file.

Example: https://github.com/elastic/kibana/blob/master/src/legacy/ui/public/flot-charts/index.js

ogupte committed:
[APM] created namespaced package `@kbn/elastic-idx`. Made available the
apm/common/idx and the corresponding babel plugin under this package.
Modified all imports of apm/common/idx to @kbn/elastic-idx.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10204/)

jasonrhodes reviewed (commented):
Trying to build this branch I see the following error:
```
   │ERROR Error: Command failed: ./bin/kibana --env.name=production --logging.json=false --optimize
   │
   │       FATAL  SyntaxError: Unexpected identifier
   │
   │
   │        log   [12:30:38.810] [info][plugins-service] Plugin initialization disabled.
   │        log   [12:30:47.514] [fatal][root] /Users/jason/__projects__/kibana/build/kibana/node_modules/@kbn/elastic-idx/src/index.ts:12
   │      interface DeepRequiredArray<T> extends Array<DeepRequired<T>> {}
   │                ^^^^^^^^^^^^^^^^^
   │
   │      SyntaxError: Unexpected identifier
```

ogupte committed:
[APM] add typescript build to the @kbn/elastic-idx package

ogupte committed:
[APM] fix idx import and linting failures

ogupte commented:
> Trying to build this branch I see the following error:
> 
> ```
>    │ERROR Error: Command failed: ./bin/kibana --env.name=production --logging.json=false --optimize
>    │
>    │       FATAL  SyntaxError: Unexpected identifier
>    │
>    │
>    │        log   [12:30:38.810] [info][plugins-service] Plugin initialization disabled.
>    │        log   [12:30:47.514] [fatal][root] /Users/jason/__projects__/kibana/build/kibana/node_modules/@kbn/elastic-idx/src/index.ts:12
>    │      interface DeepRequiredArray<T> extends Array<DeepRequired<T>> {}
>    │                ^^^^^^^^^^^^^^^^^
>    │
>    │      SyntaxError: Unexpected identifier
> ```

This should be resolved with the added build step for the typescript module

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10400/)

jasonrhodes commented:
OK cool, I pulled all the latest commits and the build got much further (it still eventually failed while making one of the OSS archive files, but it may have just run out of memory since the build takes so long, not sure).

Another thing I'm confused about: after running the build, I went into the build folder to investigate. I'm not 100% sure what I should see, but I went to a random file here: `build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/x-pack/plugins/apm/public/components/app/ErrorGroupDetails/DetailView/ErrorTabs.js` and noticed this code:
```js
const elastic_idx_1 = require("@kbn/elastic-idx");

// etc

const hasLogStacktrace = !lodash_1.isEmpty(elastic_idx_1.idx(error, _ => _.error.log.stacktrace));
```

Shouldn't that be babel-transpiled at this point? How can we confirm the new IDX babel plugin is running and converting the code?

FWIW, this is the build error I got:
```
 info [  kibana  ] Creating deb package
   │ debg calling fpm with args: [ '--force',
   │        '-t',
   │        'deb',
   │        '--package',
   │        '/Users/jason/__projects__/kibana/target/NAME-8.0.0-SNAPSHOT-ARCH.TYPE',
   │        '-s',
   │        'dir',
   │        '--name',
   │        'kibana',
   │        '--description',
   │        'Explore and visualize your Elasticsearch data',
   │        '--version',
   │        '8.0.0-SNAPSHOT',
   │        '--url',
   │        'https://www.elastic.co',
   │        '--vendor',
   │        'Elasticsearch, Inc.',
   │        '--maintainer',
   │        'Kibana Team <info@elastic.co>',
   │        '--license',
   │        'Elastic-License',
   │        '--conflicts',
   │        'kibana-oss',
   │        '--after-install',
   │        '/Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/post_install.sh',
   │        '--before-install',
   │        '/Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/pre_install.sh',
   │        '--before-remove',
   │        '/Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/pre_remove.sh',
   │        '--after-remove',
   │        '/Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/post_remove.sh',
   │        '--config-files',
   │        '/etc/kibana/kibana.yml',
   │        '--template-value',
   │        'user=kibana',
   │        '--template-value',
   │        'group=kibana',
   │        '--template-value',
   │        'optimizeDir=/usr/share/kibana/optimize',
   │        '--template-value',
   │        'configDir=/etc/kibana',
   │        '--template-value',
   │        'pluginsDir=/usr/share/kibana/plugins',
   │        '--template-value',
   │        'dataDir=/var/lib/kibana',
   │        '--exclude',
   │        'usr/share/kibana/config',
   │        '--exclude',
   │        'usr/share/kibana/data',
   │        '--architecture',
   │        'amd64',
   │        '--deb-priority',
   │        'optional',
   │        '/Users/jason/__projects__/kibana/build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/=/usr/share/kibana/',
   │        '/Users/jason/__projects__/kibana/build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/config/=/etc/kibana/',
   │        '/Users/jason/__projects__/kibana/build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/data/=/var/lib/kibana/',
   │        '/Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/service_templates/sysv/etc/=/etc/',
   │        '/Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/service_templates/systemd/etc/=/etc/' ]
   │ info $ fpm --force -t deb --package /Users/jason/__projects__/kibana/target/NAME-8.0.0-SNAPSHOT-ARCH.TYPE -s dir --name kibana --description Explore and visualize your Elasticsearch data --version 8.0.0-SNAPSHOT --url https://www.elastic.co --vendor Elasticsearch, Inc. --maintainer Kibana Team <info@elastic.co> --license Elastic-License --conflicts kibana-oss --after-install /Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/post_install.sh --before-install /Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/pre_install.sh --before-remove /Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/pre_remove.sh --after-remove /Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/package_scripts/post_remove.sh --config-files /etc/kibana/kibana.yml --template-value user=kibana --template-value group=kibana --template-value optimizeDir=/usr/share/kibana/optimize --template-value configDir=/etc/kibana --template-value pluginsDir=/usr/share/kibana/plugins --template-value dataDir=/var/lib/kibana --exclude usr/share/kibana/config --exclude usr/share/kibana/data --architecture amd64 --deb-priority optional /Users/jason/__projects__/kibana/build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/=/usr/share/kibana/ /Users/jason/__projects__/kibana/build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/config/=/etc/kibana/ /Users/jason/__projects__/kibana/build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/data/=/var/lib/kibana/ /Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/service_templates/sysv/etc/=/etc/ /Users/jason/__projects__/kibana/src/dev/build/tasks/os_packages/service_templates/systemd/etc/=/etc/
   │ERROR failure
   │ERROR Error: spawn fpm ENOENT
   │          at Process.ChildProcess._handle.onexit (internal/child_process.js:240:19)
   │          at onErrorNT (internal/child_process.js:415:16)
   │          at process._tickCallback (internal/process/next_tick.js:63:19)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

tylersmalley commented:
@jasonrhodes fpm is used to build OS specific packages (deb/rpm). I would recommend running the build with `yarn build --skip-os-packages`

ogupte committed:
[APM] make @kbn/elastic-idx/babel plugin detect idx calls as member functions from the bound scope

ogupte commented:
> OK cool, I pulled all the latest commits and the build got much further (it still eventually failed while making one of the OSS archive files, but it may have just run out of memory since the build takes so long, not sure).
> 
> Another thing I'm confused about: after running the build, I went into the build folder to investigate. I'm not 100% sure what I should see, but I went to a random file here: `build/default/kibana-8.0.0-SNAPSHOT-linux-x86_64/x-pack/plugins/apm/public/components/app/ErrorGroupDetails/DetailView/ErrorTabs.js` and noticed this code:
> 
> ```js
> const elastic_idx_1 = require("@kbn/elastic-idx");
> 
> // etc
> 
> const hasLogStacktrace = !lodash_1.isEmpty(elastic_idx_1.idx(error, _ => _.error.log.stacktrace));
> ```
> 
> Shouldn't that be babel-transpiled at this point? How can we confirm the new IDX babel plugin is running and converting the code?
> ...

While this is an intermediate step in the build, the babel plugin was not able to detect this syntax as a valid idx call, so in the bundled, production code the idx calls persisted. This is fixed in the lastest commit: while the intermediate step still produces code with the idx call, in the final step that bundles it is able to detect it as an idx call and transform it before it lands in the production build.

For instance, after running the build, take a look at the bundle file (`build/default/kibana-8.0.0-darwin-x86_64/optimize/bundles/apm.bundle.js`) notice that it has no references to the idx call as all those calls were inlined as the expanded form optional chaining.

I use this command while testing the build process, which takes less time than the full `yarn build`:

```
node scripts/build.js --skip-archives --skip-os-packages --no-oss --release --skip-node-download --verbose
```

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10471/)

jasonrhodes commented:
> For instance, after running the build, take a look at the bundle file (`build/default/kibana-8.0.0-darwin-x86_64/optimize/bundles/apm.bundle.js`) notice that it has no references to the idx call as all those calls were inlined as the expanded form optional chaining.

:clap: nice!

tylersmalley reviewed (approved):
LGTM from an Ops standpoint.

pleunv commented:
Hey guys!

Just wondering, any chance you might make this package publicly available?

sqren commented:
Hey @pleunv, 
We might eventually do that but currently we'll keep it as an internal dependency for simplicity reasons. 
For now I'd suggest you use the original package from https://github.com/facebookincubator/idx.

Btw We chose to fork it due to the issues outlined in https://github.com/facebookincubator/idx/pull/71
